tlv/varint.go

```diff
@@ -0,0 +1,109 @@
+package tlv
+
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+)
+
+// ErrCanonicalVarInt signals that the decoded varint was not minimally encoded.
+var ErrCanonicalVarInt = errors.New("decoded varint is not canonical")
+
+// ReadVarInt reads a variable length integer from r and returns it as a uint64.
+func ReadVarInt(r io.Reader, buf *[8]byte) (uint64, error) {
+	_, err := io.ReadFull(r, buf[:1])
+	if err != nil {
+		return 0, err
+	}
+	discriminant := buf[0]
+
+	var rv uint64
+	switch {
+	case discriminant < 0xfd:
+		rv = uint64(discriminant)
+
+	case discriminant == 0xfd:
+		_, err := io.ReadFull(r, buf[:2])
+		switch {
+		case err == io.EOF:
+			return 0, io.ErrUnexpectedEOF
+		case err != nil:
+			return 0, err
+		}
+		rv = uint64(binary.LittleEndian.Uint16(buf[:2]))
+
+		// The encoding is not canonical if the value could have been
+		// encoded using fewer bytes.
+		if rv < 0xfd {
+			return 0, ErrCanonicalVarInt
+		}
+
+	case discriminant == 0xfe:
+		_, err := io.ReadFull(r, buf[:4])
+		switch {
+		case err == io.EOF:
+			return 0, io.ErrUnexpectedEOF
+		case err != nil:
+			return 0, err
+		}
+		rv = uint64(binary.LittleEndian.Uint32(buf[:4]))
+
+		// The encoding is not canonical if the value could have been
+		// encoded using fewer bytes.
+		if rv < 0x10000 {
+			return 0, ErrCanonicalVarInt
+		}
+
+	default:
+		_, err := io.ReadFull(r, buf[:])
+		switch {
+		case err == io.EOF:
+			return 0, io.ErrUnexpectedEOF
+		case err != nil:
+			return 0, err
+		}
+		rv = binary.LittleEndian.Uint64(buf[:])
+
+		// The encoding is not canonical if the value could have been
+		// encoded using fewer bytes.
+		if rv < 0x100000000 {
+			return 0, ErrCanonicalVarInt
+		}
+	}
+
+	return rv, nil
+}
+
+// WriteVarInt serializes val to w using a variable number of bytes depending
+// on its value.
+func WriteVarInt(w io.Writer, val uint64, buf *[8]byte) error {
+	switch {
+	case val < 0xfd:
+		buf[0] = uint8(val)
+		_, err := w.Write(buf[:1])
+		return err
+
+	case val <= 0xffff:
+		buf[0] = uint8(0xfd)
+		binary.LittleEndian.PutUint16(buf[1:3], uint16(val))
+		_, err := w.Write(buf[:3])
+		return err
+
+	case val <= 0xffffffff:
+		buf[0] = uint8(0xfe)
+		binary.LittleEndian.PutUint32(buf[1:5], uint32(val))
+		_, err := w.Write(buf[:5])
+		return err
+
+	default:
+		buf[0] = uint8(0xff)
+		_, err := w.Write(buf[:1])
```

Comment:
Do we need a 9-byte buffer?

Reply:
i considered it but decided against it to promote cache alignment. in practice we will probably never read/write a number this big, so seemed okay to special case